File: tests/fixtures/typescript/sample-repo/src/net.ts
// Network-effect patterns.

import { raiseIfInvalid } from "./errors.js";

export async function fetchUser(id: number): Promise<unknown> {
  // expect: Net=definite, Async=definite, Throw=probable
  raiseIfInvalid(id);
  const res = await fetch(`https://api.example/users/${id}`);
  if (!res.ok) throw new Error(`status ${res.status}`);
  return res.json();
}

export async function postEvent(payload: unknown): Promise<void> {
  // expect: Net=definite, Async=definite
  await fetch("https://api.example/events", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function openWebSocket(url: string): WebSocket {
  // expect: Net=definite
  return new WebSocket(url);
}

export function makeRequestHelper(url: string): string {
  // expect: Net=possible
  // Pure body; name pattern only.
  return url.replace(/\/+$/, "");
}

declare interface RestClient {
  get(path: string): Promise<unknown>;
}

export async function fetchAllThings(client: RestClient): Promise<unknown[]> {
  // expect: Net=probable
  // `client` is annotated as a Client-typed value; methods are probable Net.
  return Promise.all([client.get("/a"), client.get("/b")]);
}

File: tests/fixtures/rust/sample-repo/src/net.rs
// Network-effect patterns.

use reqwest::Client;
use std::net::{TcpListener, TcpStream};

use crate::errors::raise_if_invalid;

pub async fn fetch_user(client: &Client, id: u64) -> Result<String, reqwest::Error> {
    // expect: Net=definite, Async=definite, Throw=probable
    raise_if_invalid(id);
    let url = format!("https://api.example/users/{id}");
    let resp = client.get(&url).send().await?;
    resp.text().await
}

pub async fn post_event(client: &Client, payload: &str) -> Result<(), reqwest::Error> {
    // expect: Net=definite, Async=definite
    client.post("https://api.example/events").body(payload.to_string()).send().await?;
    Ok(())
}

pub fn open_tcp(addr: &str) -> std::io::Result<TcpStream> {
    // expect: Net=definite
    TcpStream::connect(addr)
}

pub fn bind_listener(addr: &str) -> std::io::Result<TcpListener> {
    // expect: Net=definite
    TcpListener::bind(addr)
}

pub fn make_request_helper(url: &str) -> String {
    // expect: Net=possible
    // Pure body; name pattern only.
    url.trim_end_matches('/').to_string()
}
